trackstop.lua

```diff
@@ -236,7 +268,197 @@ function RollerOverlay:init()
+  local same_route = self.selected_stop.route_index == item.route_index
+
+  for i, stop in ipairs(item_route.stops) do
+    stop.id = i + 1
+    reset_guide_paths(stop.conditions)
+  end
+
+  if not same_route and self.selected_stop then
+    for i, stop in ipairs(routes[self.selected_stop.route_index].stops) do
+      stop.id = i + 1
+      reset_guide_paths(stop.conditions)
+    end
+  end
+end
+
+function ReorderStopsWindow:init()
+  self.selected_stop = nil
+  self:addviews{
+    widgets.Label{
+      frame={t=0,l=0},
+      view_id='hint',
+      text=SELECT_STOP_HINT,
+    },
+    widgets.List{
+      view_id='routes',
+      frame={t=2,l=1},
+      choices={},
+      on_select=function(_, item)
+        if not item then return end
+        if item.type == 'stop' then
+          local item_pos = df.global.plotinfo.hauling.routes[item.route_index].stops[item.stop_index].pos
+          dfhack.gui.revealInDwarfmodeMap(item_pos, true, true)
+        end
+      end,
+      on_submit=function(index, item)
+        self:handleStopSelection(index, item)
+      end,
+    },
+  }
+
+  self:updateList()
+end
+
+function ReorderStopsWindow:updateList()
+  local routes = df.global.plotinfo.hauling.routes
+  local choices = {}
+  local list_position = 0
+
+  if self.selected_stop then
+    self.subviews.hint:setText(SELECT_ANOTHER_STOP_HINT)
+  else
+    self.subviews.hint:setText(SELECT_STOP_HINT)
+  end
+
+  for i, route in ipairs(routes) do
+    local stops = route.stops
+    local route_name = route.name
+
+    if route_name == '' then
+      route_name = 'Route ' .. route.id
+    end
+
+    table.insert(choices, {text=route_name, type='route', route_index=i, list_position=list_position})
+    list_position = list_position + 1
+
+    for j, stop in ipairs(stops) do
+      local stop_name = stop.name
+
+      if stop_name == '' then
+        stop_name = 'Stop ' .. stop.id
+      end
+
+      if self.selected_stop and self.selected_stop.list_position == list_position then
+        stop_name = '=> ' .. stop_name
+      end
+
+      stop_name = '  ' .. stop_name
+
+      table.insert(choices, {text=stop_name, type='stop', stop_index=j, route_index=i, list_position=list_position})
+      list_position = list_position + 1
+    end
+  end
+
+  self.subviews.routes:setChoices(choices)
+end
+
+ReorderStopsModal = defclass(ReorderStopsModal, gui.ZScreenModal)
+
+ReorderStopsModal.ATTRS = {
+  focus_path = 'ReorderStops',
+}
+
+function ReorderStopsModal:init()
+  self:addviews{ReorderStopsWindow{}}
+end
+
+function ReorderStopsModal:onDismiss()
+  df.global.game.main_interface.recenter_indicator_m.x = -30000
+  df.global.game.main_interface.recenter_indicator_m.y = -30000
+  df.global.game.main_interface.recenter_indicator_m.z = -30000
+end
+
+ReorderStopsOverlay = defclass(ReorderStopsOverlay, overlay.OverlayWidget)
+ReorderStopsOverlay.ATTRS{
+  default_pos={x=6, y=6},
+  default_enabled=true,
+  viewscreens='dwarfmode/Hauling',
```

Comment:
is this specific enough, or should it disappear on hauling subscreens? We might need to extend Gui.cpp if this is the case

Reply:
It looks like it's okay on all of them except them item selection one. Will fix.